common/kvstore/src/main/java/org/apache/spark/util/kvstore/RocksDB.java

```diff
@@ -0,0 +1,436 @@
+  public void close() throws IOException {
+    synchronized (this._db) {
+      org.rocksdb.RocksDB _db = this._db.getAndSet(null);
+      if (_db == null) {
+        return;
+      }
+
+      try {
+        if (iteratorTracker != null) {
+          for (Reference<RocksDBIterator<?>> ref: iteratorTracker) {
+            RocksDBIterator<?> it = ref.get();
+            if (it != null) {
+              it.close();
+            }
+          }
+        }
+        _db.close();
+      } catch (IOException ioe) {
+        throw ioe;
+      } catch (Exception e) {
+        throw new IOException(e.getMessage(), e);
+      }
+    }
+  }
+
+  /**
+   * Closes the given iterator if the DB is still open. Trying to close a JNI RocksDB handle
+   * with a closed DB can cause JVM crashes, so this ensures that situation does not happen.
+   */
+  void closeIterator(RocksDBIterator<?> it) throws IOException {
+    notifyIteratorClosed(it);
+    synchronized (this._db) {
+      org.rocksdb.RocksDB _db = this._db.get();
+      if (_db != null) {
+        it.close();
+      }
+    }
+  }
+
+  /**
+   * Remove iterator from iterator tracker. `RocksDBIterator` calls it to notify
+   * iterator is closed.
+   */
+  void notifyIteratorClosed(RocksDBIterator<?> it) {
+    iteratorTracker.removeIf(ref -> it.equals(ref.get()));
+  }
+
+  /** Returns metadata about indices for the given type. */
+  RocksDBTypeInfo getTypeInfo(Class<?> type) throws Exception {
+    RocksDBTypeInfo ti = types.get(type);
+    if (ti == null) {
+      RocksDBTypeInfo tmp = new RocksDBTypeInfo(this, type, getTypeAlias(type));
+      ti = types.putIfAbsent(type, tmp);
+      if (ti == null) {
+        ti = tmp;
+      }
+    }
+    return ti;
+  }
+
+  /**
+   * Try to avoid use-after close since that has the tendency of crashing the JVM. This doesn't
+   * prevent methods that retrieved the instance from using it after close, but hopefully will
+   * catch most cases; otherwise, we'll need some kind of locking.
+   */
+  org.rocksdb.RocksDB db() {
+    org.rocksdb.RocksDB _db = this._db.get();
+    if (_db == null) {
+      throw new IllegalStateException("DB is closed.");
+    }
+    return _db;
+  }
+
+  private byte[] getTypeAlias(Class<?> klass) throws Exception {
+    byte[] alias = typeAliases.get(klass.getName());
+    if (alias == null) {
+      synchronized (typeAliases) {
+        byte[] tmp = String.valueOf(typeAliases.size()).getBytes(UTF_8);
+        alias = typeAliases.putIfAbsent(klass.getName(), tmp);
+        if (alias == null) {
+          alias = tmp;
+          put(TYPE_ALIASES_KEY, new TypeAliases(typeAliases));
+        }
+      }
+    }
+    return alias;
+  }
+
+  /** Needs to be public for Jackson. */
+  public static class TypeAliases {
+
+    public Map<String, byte[]> aliases;
+
+    TypeAliases(Map<String, byte[]> aliases) {
+      this.aliases = aliases;
+    }
+
+    TypeAliases() {
+      this(null);
+    }
+
+  }
+
+  private static class PrefixCache {
```

Comment:
`PrefixCache` is repeatedly defined here because this pr does not want to introduce too many refactoring work, right?

Reply:
Yes, exactly, @LuciferYang . You can do `diff LevelDB.java RocksDB.java` to see the difference.
We are going to remove LevelDB in the future.

Reply:
OK ~